docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+
+*Banning a peer in case of duplicate transactions*
+
+Currently, a peer can send the same valid (or invalid) transaction [multiple times](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L247). Peers do not 
+gossip transactions to peers that have sent them that same transaction. But there is no check on whether 
+a node has alrady sent the same transaction to this peer before. There is also no check whether the transaction
+that is being gossiped was valid or not (assumint that invalid transactions could become valid).
+The transaction broadcast logic simply loops through the mempool and tries to send the transactions currently in the pool. 
+
+If we want to ban peers based on duplicate transactions, we should either add additional checks for the cases above, or 
+not ban peers for this behaviour at the moment. It would be useful to gather metrics on how often a peer gossips the same 
+transaction and whether this is cause of significant traffic. 
+
+#### **Banning in case of ResponseCheckTx.code > 1**
+
+If a transaction fails since it could never have been valid, `CheckTx` returns a `ResponseCheckTx.code` 
+value greater than 1. In this case, the peer should be disconnected from and banned immediately without keeping count on how often 
+this has happened. 
+
+The question is whether this transaction should be kept track of in the cache? We can still store it in 
+the cache so that we don't run `CheckTx` on it again, but if this peer is immediately banned, maybe there is no need
+to store its information.
+
+Now, if we want to differentiate further reasons of why this transaction is sent to a node (whether it is a sign of malice or not),
+we might need more information on the actual reason for rejection. This could be done by an additional set of response codes provided by the application. 
+
+### 2. Choosing the peer to ban
+
+Each transaction gossiped contains the ID of the peer that sent that transaction. Upon receiving a transaction,
+a node saves the peer ID of the peer(s) that have sent it. As each peer had to have
+run `CheckTx` on this transaction before adding it to its own mempool, we can assume this peer
+can be held accountable for the validity of transactions it gossips. Invalid transactions are kept only in the mempool
+cache and thus not gossiped. 
+
+*Transactions received by users*
+
+For transactions submitted via `boradcastTxCommit`, the `SenderID` field is empty. 
+
+**Note**  Do we have mechanisms in place to handle cases when `broadcastTxCommit` submits
+failing transactions (can this be a form of attack)?
+
+### 3. Attack scenarios
+
+While an attack by simply banning peers on faling `CheckTx` is hard to imagine, as the incentive for doing so is not clear, there are considerations with regards to the current mempool gossip implementation.
```

Comment:
The question is whether this can be used to play dirty. For instance, when I ban a peer using `CheckTx` although the transactions are fine, can I harm the other or gain advantage?

Reply:
In the current proposal, the ban is only from the point of view of one node. So if this node bans a peer, it is only him that looses connectivity while other peers have no information about this. However, I am also not sure what happens if a node bans peers it does not like for some reason? Somehow, as long as 2/3 of the network is honest and the peer you are banning has other connections, this should not impact the correctness. Is there a scenario this malicious node is the only connection of a peer (maybe if its a seed node). Thus when it bans it, this node can never join the network as it does not find out about other nodes? But I see this as a possible scenario if a node for any reason drops connections in any other part of Tendermint.